### DISCUSSION TITLE
Restart and Recover Segmentation Fault when Mesh is Modified

### DISCUSSION THREAD

**Discussion by czadeh** [idaholab/moose#25264] [Q&A Meshing]
Hi,

I have tried using restart and recover on a simulation where I modified the mesh in the middle of the simulation by using CoupledVarThresholdElementSubdomainModifier. 

I get a segmentation fault when I try to do so. It says PID 5709, exit code 9, Segmentation Fault (signal 11). When I try to recover on an earlier checkpoint before the mesh is modified, it works.

Is there any way around this problem?

Thanks,
Cameron

**Comment by GiudGiud** [comment]
Hello

We are in the midst of reworking this system, it should be more resilient to that soon.

For now you can try to use the SolutionUserObject / SolutionAux to restart instead. SolutionUserObject will be more capable of handling domain changes.

Guillaume

**Reply by czadeh** [reply]
Good to hear that it will be reworked soon.

Is there a way to get SolutionUserObject / SolutionAux to import a stateful material property from a previous simulation?

**Reply by GiudGiud** [reply]
ah no I dont think so.
@loganharbour we have no alternative than Checkpoint for stateful matprop right now do we?

**Reply by loganharbour** [reply]
No, we don't. And that likely will not change

**Reply by maxnezdyur** [reply]
This problem may be related to https://github.com/idaholab/moose/discussions/22330. If you are moving elements around from a subdomain that had variables to another that doesn't have those same variables. 
> We are in the midst of reworking this system, it should be more resilient to that soon.

I don't think this particular problem will be fixed, but I think there may be an error that get's thrown instead of segfaulting.
@loganharbour  There is a MWE in that issue.

**Comment by loganharbour** [comment]
I can try to make some sense of a backtrace if you can get us one.

See https://mooseframework.inl.gov/application_development/debugging.html for more information. Once you've set a breakpoint, ran, and it has crashed, share the result of running `bt` in the debugger.

If you have an input that you're willing to share, I might be able to take a look there as well.

**Reply by czadeh** [reply]
I'll work on recreating the problem in a minimal input file and get that to you soon.

I tried debugging on my own but I couldn't get any good information out of it.

**Reply by GiudGiud** [reply]
ah forgot to pitch in earlier.
 
We are down for the backtrace, but at this point I believe the way forward is with a SolutionUserObject and SOlutionIC/Aux to set variables

this will enable non-matching meshes

**Reply by czadeh** [reply]
So I made some discoveries.

I created a minimal input file for testing and found that I could restart from a checkpoint without issue. Here is the file for reference. I used `--recover` at checkpoint 0008, after the mesh had changed, and it works.

```
filebase = "SubdomainTestSimple"
[Problem]
    kernel_coverage_check = false
    type = FEProblem
[]
[Mesh]
    [gen]
        type = GeneratedMeshGenerator
        dim = 3
        nx = 20
        ny = 20
        nz = 6
        xmax = 2e-3       # meters
        ymax = 2e-3
        zmax = 0.6e-3
    []
    [active_domain] #the initially active domain
        type = SubdomainBoundingBoxGenerator
        input = 'gen'
        block_id = 1
        bottom_left = '0 0 0'
        top_right = '2e-3 2e-3 0.4e-3'
    []
    [inactive_domain] #the initially inactive domain
        type = SubdomainBoundingBoxGenerator
        input = active_domain
        block_id = 2
        bottom_left = '0 0 0.4e-3'
        top_right = '2e-3 2e-3 0.6e-3'
    []
    [sideset_top]
        input = inactive_domain
        type = SideSetsAroundSubdomainGenerator
        normal = '0 0 1'
        block = 1
        new_boundary = 'topLayer' # top boundary of active domain
    []
[]
[AuxVariables]
    [layerHeight]
        outputs = none
    []
    [layerZ]
        outputs = none
    []
[]
[AuxKernels]
    [layerHeightKernel]
        type = FunctionAux
        variable = layerHeight
        function = path_z
        execute_on = 'TIMESTEP_BEGIN'
    []
    [layerZKernel]
        type = ParsedAux
        variable = layerZ
        use_xyzt = true
        coupled_variables = 'layerHeight'
        expression = 'z - layerHeight'
        execute_on = 'TIMESTEP_BEGIN'
    []
[]
[UserObjects]
    [addLayer] #subdomain modifier
        type = CoupledVarThresholdElementSubdomainModifier
        coupled_var = 'layerZ' #the z height of the active layers
        criterion_type = BELOW
        threshold = 0
        subdomain_id = 1
        moving_boundary_name = topLayer
        execute_on = 'TIMESTEP_BEGIN'
    []
[]
[Materials]
    [Generic]
        type = ADGenericConstantMaterial
        prop_names = 'k rho cpT'
        prop_values = '0.3 1e6 2'
    []
    [dummy_mat_inactive]
        type = GenericConstantMaterial
        prop_names = 'dummy_mat'
        prop_values = '0.0'
        block = 2
    []
[]
[Functions]
    [path_z]
        type = ParsedFunction
        expression = '4e-4 + 1e-4*(t>5e-4)'
    []
[]
[Variables]
    [T]
        block = 1
        initial_condition = 440
    []
[]
[Kernels]
    [conduction]
        block = 1
        type = ADHeatConduction
        variable = T
        thermal_conductivity = k
    []
    [conduction_time_derivative]
        block = 1
        type = ADHeatConductionTimeDerivative
        density_name = rho
        specific_heat = cpT
        variable = T
    []
[]
[BCs]
    [top]
        type = ADConvectiveHeatFluxBC
        variable = T
        boundary = topLayer
        T_infinity = 400
        heat_transfer_coefficient = 15 # W /(m2*K)
    []
[]
[Preconditioning]
    [full]
        type = SMP
        full = true
    []
[]
[Executioner]
    type = Transient
    solve_type = PJFNK
    petsc_options_iname = '-ksp_gmres_restart -pc_type -pc_hypre_type -pc_hypre_boomeramg_max_iter -pc_hypre_boomeramg_strong_threshold'
    petsc_options_value = '201 hypre boomeramg 8 0.7'
    nl_rel_tol = 1e-10
    nl_abs_tol = 1e-12
    [TimeSteppers]
        [Constant]
            type = ConstantDT
            dt = 1e-4
        []
    []
    end_time = 1e-3
[]
[Outputs]
    [ex]
        type = Exodus
        file_base = ./paramTests/out/${filebase}_out
    [] 
    [check]
        type = Checkpoint
        file_base = ./paramTests/out/${filebase}_check
        interval = 4
    [] 
[]
```

I narrowed down the problem to see that it only happens when I have my custom stateful material property enabled, in which recovering from a checkpoint leads to a seg fault. I started gdb with my app's dbg file and with --recover and checkpoint as part of the command. This is the error message that I got:

```
Assertion `prop_size == v.size()' failed
Loading MaterialProperties data into mis-sized target
at /home/cameron/projects/moose/framework/src/materials/MaterialProperty.C, line 39
Stack frames: 20
0: libMesh::print_trace(std::ostream&)
1: dataLoad(std::istream&, MaterialProperties&, void*)
2: /home/cameron/projects/moose/framework/libmoose-dbg.so.0(+0x1ba2977) [0x7ffff51a2977]
3: /home/cameron/projects/moose/framework/libmoose-dbg.so.0(+0x1b93c3e) [0x7ffff5193c3e]
4: /home/cameron/projects/moose/framework/libmoose-dbg.so.0(+0x1b84aa9) [0x7ffff5184aa9]
5: /home/cameron/projects/moose/framework/libmoose-dbg.so.0(+0x1b7827c) [0x7ffff517827c]
6: dataLoad(std::istream&, MaterialPropertyStorage&, void*)
7: /home/cameron/projects/moose/framework/libmoose-dbg.so.0(+0x270b5e8) [0x7ffff5d0b5e8]
8: /home/cameron/projects/moose/framework/libmoose-dbg.so.0(+0x2706d40) [0x7ffff5d06d40]
9: RestartableDataIO::deserializeRestartableData(std::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::unique_ptr<RestartableDataValue, std::default_delete<RestartableDataValue> >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::unique_ptr<RestartableDataValue, std::default_delete<RestartableDataValue> > > > > const&, std::istream&, std::unordered_set<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&)
10: RestartableDataIO::readRestartableData(std::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::unique_ptr<RestartableDataValue, std::default_delete<RestartableDataValue> >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::unique_ptr<RestartableDataValue, std::default_delete<RestartableDataValue> > > > > const&, std::unordered_set<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, unsigned int)
11: RestartableDataIO::readRestartableData(std::vector<std::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::unique_ptr<RestartableDataValue, std::default_delete<RestartableDataValue> >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::unique_ptr<RestartableDataValue, std::default_delete<RestartableDataValue> > > > >, std::allocator<std::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::unique_ptr<RestartableDataValue, std::default_delete<RestartableDataValue> >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::unique_ptr<RestartableDataValue, std::default_delete<RestartableDataValue> > > > > > > const&, std::unordered_set<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&)
12: FEProblemBase::initialSetup()
13: Transient::init()
14: MooseApp::executeExecutioner()
15: MooseApp::run()
16: main
17: /usr/lib/libc.so.6(+0x27cd0) [0x7fffea027cd0]
18: __libc_start_main
19: /home/cameron/projects/turtle/turtle-dbg(+0x3079) [0x555555557079]
[0] /home/cameron/projects/moose/framework/src/materials/MaterialProperty.C, line 39, compiled Aug 20 2023 at 18:05:22

Thread 1 "turtle-dbg" hit Breakpoint 1, 0x00007fffea4b0820 in PMPI_Abort () from /home/cameron/mambaforge3/envs/moose/lib/libmpi.so.12
```

I don't know what to make of this.

@GiudGiud I have used SolutionUserObject/SolutionAux with FunctionIC before, but I haven't found a way to use it for material properties. Is this possible? Is SolutionIC different? I see in the documentation that SolutionIC is meant for variables, but there is an optional parameter for prop_getter_suffix, will that enable its usage for material properties?

Thanks

**Reply by GiudGiud** [reply]
Stateful material properties are indeed one of the complex part of the restart. Seemingly they struggle more here because materials are tied to subdomains, so if subdomains change the code has to be very flexible to understand what's going on

SolutionAux/IC are both for variables. 
If you wanted to approximate the material, you could convert it to a variable, save it to a solution file, then load that, then rebuild the material using a ParsedMaterial from the variable. This will be approximate but it will work

**Comment by loganharbour** [comment]
The new restart/recover changes are in with quite a bit of error checking.

If you have the time - please update moose (will require a libmesh rebuild or a mamba/conda update) and try again. You'll need to re-run your original problem to produce the checkpoint as well. Glad to help now that the restart/recover changes are mostly finalized, but I'd like to make sure you're still hitting the same error.

**Reply by czadeh** [reply]
I am interested to try.

I performed `mamba update --all`, `git fetch origin`, and `git rebase origin/master`. Is this the right way to perform updates?

I saw this post (https://github.com/idaholab/moose/discussions/25358) and wasn't sure whether it was still the case to update this way. I did not update like that this time.

I performed `make -j 16` in moose/test and then `./run_tests -j 16` and only 2 tests failed: 

```
meshgenerators/meta_data_store.test_meta_data_with_use_split ....................... [skipped dependency] SKIP
meshgenerators/meta_data_store.pre_split_mesh ............................... FAILED (EXPECTED OUTPUT MISSING)
mesh/mesh_only.mesh_only_checkpoint ............................................. [min_cpus=3] FAILED (ERRMSG)
```

Are these new tests added for the new functionality? They seem to be related to the mesh system and checkpoints.

**Reply by milljm** [reply]
I cannot speak for the failing tests, but I wanted to comment:

Once you perform https://github.com/idaholab/moose/discussions/25358, `mamba update --all` routine should work as before. 

Due to the uncertainty of it, we are looking into a better solution that will alert folks when there is a mismatch.

You can always tell if you're using the right "version" of our packages, based on the version of MOOSE you have checked out by running the versioner tool:

```bash
❯ mamba activate moose
❯ cd moose
❯ scripts/versioner.py libmesh --yaml | grep install
  install: moose-libmesh=2023.09.06=build_0
```
Armed with that information you now know what version of `moose-libmesh` is required. And can compare it to what is installed:
```bash
❯ mamba list | grep moose-libmesh
moose-libmesh       2023.09.06      build_0     https://conda.software.inl.gov/public
```

**Reply by czadeh** [reply]
@milljm Thanks, I tried those commands and got the same results as you so everything is up to date.

**Reply by czadeh** [reply]
I tried recovering from a checkpoint when the solver failed to converge in a problem where I have stateful properties and a changing mesh.

I found a parameter that I supplied through a CSV which was causing the issue. I changed the CSV and tried to recover from a checkpoint.

The good news is that I was able to recover and continue the simulation from the checkpoint. Good to see that it will work for this case.

The bad news is that the change that I made to my CSV did not affect the way that the recovered simulation ran. It still failed to solve at the same point. When I ran the input file from the beginning again, things went smoothly. 

Is recover not the right tool for me to use in this situation?

**Reply by GiudGiud** [reply]
So you can go further when running from the beginning of the simulation than when:
- running for part of it, stopping, changing a CSV value, then restarting from checkpoint?
and
- running for part of it, stopping, then restarting from checkpoint?

**Reply by czadeh** [reply]
> So you can go further when running from the beginning of the simulation than when:
> 
> * running for part of it, stopping, changing a CSV value, then restarting from checkpoint?

Yes, I can go further when I change a problematic parameter in the CSV.

> * running for part of it, stopping, then restarting from checkpoint?

No difference here.
